charon/src/ast/types_utils.rs

```diff
@@ -518,3 +525,118 @@ impl<V> std::ops::DerefMut for VisitInsideTy<V> {
+        match item.downcast_mut::<Region>() {
+            Some(r) => {
+                match r {
+                    Region::BVar(db, id) => {
+                        if db.index == self.current_level {
+                            *r = self.region_subst.get(*id).unwrap().clone()
+                        }
+                    }
+                    _ => (),
+                };
+                true
+            }
+            None => false,
+        }
+    }
+
+    /// Returns [true] if the item is a binder (we use this to skip some checks)
+    fn try_subst_in_ty(&mut self, item: &mut dyn std::any::Any) -> bool {
+        match item.downcast_mut::<Ty>() {
+            Some(ty) => {
+                match ty.kind() {
+                    TyKind::TypeVar(id) => *ty = self.ty_subst.get(*id).unwrap().clone(),
+                    _ => (),
+                };
+                true
+            }
+            None => false,
+        }
+    }
+
+    fn try_subst_in_const_generic(&mut self, item: &mut dyn std::any::Any) -> bool {
+        match item.downcast_mut::<ConstGeneric>() {
+            Some(cg) => {
+                match cg {
+                    ConstGeneric::Var(id) => *cg = self.cg_subst.get(*id).unwrap().clone(),
+                    _ => (),
+                };
+                true
+            }
+            None => false,
+        }
+    }
+
+    fn try_subst_in_trait_ref(&mut self, item: &mut dyn std::any::Any) -> bool {
+        match item.downcast_mut::<TraitRef>() {
+            Some(tr) => {
+                match &mut tr.kind {
+                    TraitRefKind::Clause(id) => *tr = self.trait_subst.get(*id).unwrap().clone(),
+                    _ => (),
+                };
+                true
+            }
+            None => false,
+        }
+    }
+}
+
+impl VisitorMut for Subst {
+    fn visit(&mut self, item: &mut dyn std::any::Any, event: Event) {
+        // Check if this is a binder
+        let enters = matches!(event, Event::Enter);
+        if self.try_enter_or_exit_binder(item, enters) {
+        }
+        // Other cases: we only do something upon entering
+        else if enters {
+            if self.try_subst_in_region(item) {
+            } else if self.try_subst_in_ty(item) {
+            } else if self.try_subst_in_const_generic(item) {
+            } else {
+                let _ = self.try_subst_in_trait_ref(item);
+            }
+        }
+    }
+}
+
+impl Ty {
+    pub fn substitute(&mut self, generics: &GenericArgs) {
+        let mut subst = Subst {
+            current_level: 0,
+            region_subst: Vector::from_iter(generics.regions.iter().cloned()),
+            ty_subst: Vector::from_iter(generics.types.iter().cloned()),
+            cg_subst: Vector::from_iter(generics.const_generics.iter().cloned()),
+            trait_subst: Vector::from_iter(generics.trait_refs.iter().cloned()),
+        };
+        self.drive_mut(&mut subst);
```

Comment:
This will not recurse deeply into types, you need to wrap `subst` in `Ty::visit_inside`